semtypes/src/main/java/io/ballerina/semtype/Core.java

```diff
@@ -69,12 +71,61 @@ public static SemType union(SemType t1, SemType t2) {
         if (t1 instanceof UniformTypeBitSet) {
             if (t2 instanceof UniformTypeBitSet) {
                 return UniformTypeBitSet.from(((UniformTypeBitSet) t1).bitset | ((UniformTypeBitSet) t2).bitset);
+            } else {
+                ComplexSemType complexT2 = (ComplexSemType) t2;
+                all2 = complexT2.all;
+                some2 = complexT2.some;
             }
+            all1 = (UniformTypeBitSet) t1;
+            some1 = UniformTypeBitSet.from(0);
         } else {
+            ComplexSemType complexT1 = (ComplexSemType) t1;
+            all1 = complexT1.all;
+            some1 = complexT1.all;
+            if (t2 instanceof UniformTypeBitSet) {
+                all2 = ((UniformTypeBitSet) t2);
+                some2 = UniformTypeBitSet.from(0);
+            } else {
+                ComplexSemType complexT2 = (ComplexSemType) t2;
+                all2 = complexT2.all;
+                some2 = complexT2.some;
+            }
+        }
 
+        UniformTypeBitSet all = UniformTypeBitSet.from(all1.bitset | all2.bitset);
+        UniformTypeBitSet some = UniformTypeBitSet.from((some1.bitset | some2.bitset) & ~all.bitset);
+        if (some.bitset == 0) {
+            return PredefinedType.uniformTypeUnion(all.bitset);
         }
 
-        throw new AssertionError("Not Implemented");
+        List<UniformSubtype> subtypes = new ArrayList<>();
+
+        for (SubtypePair pair : new SubtypePairs(t1, t2, some)) {
+            UniformTypeCode code = pair.uniformTypeCode;
+            SubtypeData data1 = pair.subtypeData1;
+            SubtypeData data2 = pair.subtypeData2;
+
+            SubtypeData data;
+            if (data1 == null) {
+                data = (SubtypeData) data2; // // [from original impl] if they are both null, something's gone wrong
+            } else if (data2 == null) {
+                data = data1;
+            } else {
+                data = OpsTable.OPS[code.code].union(data1, data2);
```

Comment:
Is there a particular reason a function object is used for this in nballerina? If needed we can maybe represent a similar code using a method reference variable.

Reply:
Not sure what you meant by `function object`. The reason for having this table of function pointers is to share common behavior as Ballerina objects don't have implementation inheritance.

Reply:
in nballerina 
`var intersect = ops[code].intersect; 
 data = intersect(data1, data2);` 
 is used instead of directly using `data = ops[code].intersect(data1, data2);`
 I was wondering whether it was a ballerina limitation or a design choice

Reply:
Yah, according to the spec, method call expression only work with objects (object methods and object fields) and for langlib function invocation. 
https://ballerina.io/ballerina-spec/spec.html#section_6.16